roundRobin v2.1.0 fails where v2.0.4 does not

Apologies for the horrible report, but I'm away and wanted to bang this out quickly instead of waiting a couple weeks and probably forgetting it until next I updated.

Updating oxy/v2 from v2.0.4 to v2.1.0 causes a test of a simple roundrobin failover to fail. The text creates handlerFuncs that count their requests, upserts them into the roundrobin, CLOSES one of them, runs 10 GET requests on the roundrobin, and then checks to make sure the open server saw 10 request, and the closed server saw 0. The same loop that tuns the 10 GETs also ensures HTTP 200 is returned for each request.

The same tests against v2.1.0 have the 10 GETs returning 200, however the handlerFunc from the open server only saw 5 requests, not all 10. Shell output from the test against both versions below:

```console
m@mlap2:~/Documents/workspace/go-jar/xt$ go get github.com/vulcand/oxy/v2@v2.0.4
go: downgraded github.com/vulcand/oxy/v2 v2.1.0 => v2.0.4
m@mlap2:~/Documents/workspace/go-jar/xt$ go test -v
=== RUN   TestPoolRoundRobinFailWell

  When a two-member roundrobin is created with a buffer, and one "crashes", the failover is proper ✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔


17 total assertions

--- PASS: TestPoolRoundRobinFailWell (0.00s)
PASS
ok  	github.com/cognusion/go-jar/xt	0.006s
```


```console
m@mlap2:~/Documents/workspace/go-jar/xt$ go get github.com/vulcand/oxy/v2@v2.1.0
go: upgraded github.com/vulcand/oxy/v2 v2.0.4 => v2.1.0
m@mlap2:~/Documents/workspace/go-jar/xt$ go test -v
=== RUN   TestPoolRoundRobinFailWell

  When a two-member roundrobin is created with a buffer, and one "crashes", the failover is proper ✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✘✔


Failures:

  * /home/m/Documents/workspace/go-jar/xt/xt_test.go 
  Line 63:
  Expected: 10
  Actual:   5
  (Should equal)!


17 total assertions

--- FAIL: TestPoolRoundRobinFailWell (0.00s)
FAIL
exit status 1
FAIL	github.com/cognusion/go-jar/xt	0.005s

```

I have not dug into the details of this yet, but it's a pretty raw oxy test with very little room for "my code" to be a problem. Test code follows:

```go
import (
	. "github.com/smartystreets/goconvey/convey"
	"github.com/vulcand/oxy/v2/buffer"
	"github.com/vulcand/oxy/v2/forward"
	"github.com/vulcand/oxy/v2/roundrobin"

	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
)

func TestPoolRoundRobinFailWell(t *testing.T) {
	req := httptest.NewRequest("GET", "/", nil)

	Convey("When a two-member roundrobin is created with a buffer, and one \"crashes\", the failover is proper", t, FailureContinues, func(c C) {

		rr := httptest.NewRecorder()

		oneCount := 0
		one := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			oneCount++
			w.Write([]byte("Ok"))
		})
		oneServer := httptest.NewServer(one)
		defer oneServer.Close()
		oneURL, err := url.Parse(oneServer.URL)
		So(err, ShouldBeNil)

		twoCount := 0
		two := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			twoCount++
			w.Write([]byte("Ok"))
		})
		twoServer := httptest.NewServer(two)
		// defer twoServer.Close()
		twoURL, err := url.Parse(twoServer.URL)
		So(err, ShouldBeNil)
		// explicit close now
		twoServer.Close()

		fwd := forward.New(false)
		lb, err := roundrobin.New(fwd)
		So(err, ShouldBeNil)

		lb.UpsertServer(oneURL)
		lb.UpsertServer(twoURL)

		buff, err := buffer.New(lb, buffer.Retry(fmt.Sprintf("IsNetworkError() && Attempts() < %d", 2)))
		So(err, ShouldBeNil)

		okCount := 0
		for range 10 {
			buff.ServeHTTP(rr, req)
			So(rr.Code, ShouldEqual, http.StatusOK)
			okCount++
		}

		So(okCount, ShouldEqual, 10)
		So(oneCount, ShouldEqual, 10)
		So(twoCount, ShouldEqual, 0)
	})
}

```

--- a/buffer/buffer.go
+++ b/buffer/buffer.go
@@ -250,7 +250,7 @@ func (b *Buffer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 			reader = rdr
 		}
 
-		if body == nil || (b.retryPredicate == nil || attempt > DefaultMaxRetryAttempts) ||
+		if b.retryPredicate == nil || attempt > DefaultMaxRetryAttempts ||
 			!b.retryPredicate(&context{r: req, attempt: attempt, responseCode: bw.code}) {
 			utils.CopyHeaders(w.Header(), bw.Header())
 			w.WriteHeader(bw.code)

--- a/roundrobin/rr_test.go
+++ b/roundrobin/rr_test.go
@@ -1,12 +1,15 @@
 package roundrobin
 
 import (
+	"fmt"
 	"net/http"
 	"net/http/httptest"
+	"sync/atomic"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"github.com/vulcand/oxy/v2/buffer"
 	"github.com/vulcand/oxy/v2/forward"
 	"github.com/vulcand/oxy/v2/testutils"
 	"github.com/vulcand/oxy/v2/utils"
@@ -201,13 +204,55 @@ func TestRoundRobinRequestRewriteListener(t *testing.T) {
 	assert.NotNil(t, lb.requestRewriteListener)
 }
 
-func seq(t *testing.T, url string, repeat int) []string {
+func TestRoundRobinFailWell(t *testing.T) {
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+
+	rr := httptest.NewRecorder()
+
+	var oneCount atomic.Int32
+
+	oneServer := testutils.NewResponderWithCount(&oneCount)
+
+	t.Cleanup(func() { oneServer.Close() })
+
+	var twoCount atomic.Int32
+
+	twoServer := testutils.NewResponderWithCount(&twoCount)
+
+	// explicit close now
+	twoServer.Close()
+
+	lb, err := New(forward.New(false))
+	require.NoError(t, err)
+
+	require.NoError(t, lb.UpsertServer(testutils.MustParseRequestURI(oneServer.URL)))
+	require.NoError(t, lb.UpsertServer(testutils.MustParseRequestURI(twoServer.URL)))
+
+	buff, err := buffer.New(lb, buffer.Retry(fmt.Sprintf("IsNetworkError() && Attempts() < %d", 2)))
+	require.NoError(t, err)
+
+	okCount := 0
+
+	for range 10 {
+		buff.ServeHTTP(rr, req)
+
+		assert.Equal(t, http.StatusOK, rr.Code)
+
+		okCount++
+	}
+
+	assert.Equal(t, 10, okCount)
+	assert.Equal(t, int32(10), oneCount.Load())
+	assert.Zero(t, twoCount.Load())
+}
+
+func seq(t *testing.T, uri string, repeat int) []string {
 	t.Helper()
 
 	var out []string
 
 	for range repeat {
-		_, body, err := testutils.Get(url)
+		_, body, err := testutils.Get(uri)
 		require.NoError(t, err)
 
 		out = append(out, string(body))

--- a/testutils/utils.go
+++ b/testutils/utils.go
@@ -8,6 +8,7 @@ import (
 	"net/http/httptest"
 	"net/url"
 	"strings"
+	"sync/atomic"
 	"testing"
 
 	"github.com/vulcand/oxy/v2/internal/holsterv4/clock"
@@ -19,7 +20,7 @@ func NewHandler(handler http.HandlerFunc) *httptest.Server {
 	return httptest.NewServer(handler)
 }
 
-// NewResponder creates a new Server with response.
+// NewResponder creates a new Server with a response.
 func NewResponder(t *testing.T, response string) *httptest.Server {
 	t.Helper()
 
@@ -30,6 +31,15 @@ func NewResponder(t *testing.T, response string) *httptest.Server {
 	t.Cleanup(server.Close)
 
 	return server
+}
+
+// NewResponderWithCount creates a new Server with a call counter.
+func NewResponderWithCount(count *atomic.Int32) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		count.Add(1)
+
+		_, _ = w.Write([]byte("Ok"))
+	}))
 }
 
 // MustParseRequestURI is the version of url.ParseRequestURI that panics if incorrect, helpful to shorten the tests.

